spec/01-unit/01-db/01-schema/01-schema_spec.lua

```diff
@@ -2989,4 +3015,249 @@ describe("schema", function()
+            name = {
+              type = "string"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "name" },
+            func = function(name)
+              return { name = name:upper() }
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity)
+
+      assert.truthy(transformed_entity)
+      assert.equal("TEST1", transformed_entity.name)
+    end)
+
+    it("transforms fields with input table", function()
+      local test_schema = {
+        name = "test",
+        fields = {
+          {
+            name = {
+              type = "string"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "name" },
+            func = function(name)
+              return { name = name:upper() }
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+      local input = { name = "we have a value" }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity, input)
+
+      assert.truthy(transformed_entity)
+      assert.equal("TEST1", transformed_entity.name)
+    end)
+
+    it("skips transformation when none of input matches", function()
+      local test_schema = {
+        name = "test",
+        fields = {
+          {
+            name = {
+              type = "string"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "non_existent" },
+            func = function(non_existent)
+              return { name = non_existent:upper() }
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity)
+
+      assert.truthy(transformed_entity)
+      assert.equal("test1", transformed_entity.name)
+    end)
+
+    it("skips transformation when none of input matches using input table", function()
+      local test_schema = {
+        name = "test",
+        fields = {
+          {
+            name = {
+              type = "string"
+            },
+          },
+        },
+        transformations = {
+          {
+            input = { "name" },
+            func = function(non_existent)
+              return { name = non_existent:upper() }
+            end,
+          },
+        },
+      }
+      local entity = { name = "test1" }
+      local input = { name = nil }
+
+      local TestEntities = Schema.new(test_schema)
+      local transformed_entity, _ = TestEntities:transform(entity, input)
+
+      assert.truthy(transformed_entity)
+      assert.equal("test1", transformed_entity.name)
+    end)
+
+
+    it("transforms fields with multiple transformations", function()
```

Comment:
I think we need one more test which ensures that a transformation can modify *other fields* besides the input ones. For example:

```
          {
            input = { "name" },
            func = function(name)
              return { slug = to_slug(name) }
            end,
          },

```

This would use `name` to modify the `slug` field. Is this allowed, or disallowed?

Reply:
This leads to interesting question also... should we validate again? Or at least check that it only contains right fields? Though, I think can skip it for now and just trust the transformation writer doing the right thing?

Reply:
The test was added that checks that it can modify any field. In fact it can even introduce new fields. We don't validate after transformations (do not know if we should),

Reply:
We should not validate again, because in some use-cases transformations can intentionally produce data that is non-conforming as an input.

Since transformations are part of schema definitions and are written by the Kong/plugin developer persona and not the Kong operator persona, there's a greater level of "they should know what they're doing" expected here. In the end, since this is an arbitrary Lua function, nothing prevents them from doing anything silly (like stuff with side-effects which would break the idempotency of schema validation, or even `os.execute("sudo rm -rf /*")` :stuck_out_tongue_winking_eye: ).

> In fact it can even introduce new fields.

This would fail at the database level later, and I think it's fine, because it's a programmer error that would explode quite visibly on testing. It doesn't seem worth adding runtime checks for this just to give a different error message.